rootston/output.c

```diff
@@ -37,76 +52,56 @@ static void rotate_child_position(double *sx, double *sy, double sw, double sh,
 		struct roots_desktop *desktop, struct wlr_output *wlr_output,
 		struct timespec *when, double lx, double ly, float rotation) {
-	if (wlr_surface_has_buffer(surface)) {
-		int width = surface->current->width;
-		int height = surface->current->height;
-		int render_width = width * wlr_output->scale;
-		int render_height = height * wlr_output->scale;
-		double ox = lx, oy = ly;
-		wlr_output_layout_output_coords(desktop->layout, wlr_output, &ox, &oy);
-		ox *= wlr_output->scale;
-		oy *= wlr_output->scale;
-
-		if (wlr_output_layout_intersects(desktop->layout, wlr_output,
-				lx, ly, lx + render_width, ly + render_height)) {
-			float matrix[16];
-
-			float translate_origin[16];
-			wlr_matrix_translate(&translate_origin,
-				(int)ox + render_width / 2, (int)oy + render_height / 2, 0);
-
-			float rotate[16];
-			wlr_matrix_rotate(&rotate, rotation);
-
-			float translate_center[16];
-			wlr_matrix_translate(&translate_center, -render_width / 2,
-				-render_height / 2, 0);
-
-			float scale[16];
-			wlr_matrix_scale(&scale, render_width, render_height, 1);
-
-			float transform[16];
-			wlr_matrix_mul(&translate_origin, &rotate, &transform);
-			wlr_matrix_mul(&transform, &translate_center, &transform);
-			wlr_matrix_mul(&transform, &scale, &transform);
-			wlr_matrix_mul(&wlr_output->transform_matrix, &transform, &matrix);
-
-			wlr_render_with_matrix(desktop->server->renderer, surface->texture,
-				&matrix);
-
-			struct wlr_frame_callback *cb, *cnext;
-			wl_list_for_each_safe(cb, cnext,
-					&surface->current->frame_callback_list, link) {
-				wl_callback_send_done(cb->resource, timespec_to_msec(when));
-				wl_resource_destroy(cb->resource);
-			}
-		}
+	if (!wlr_surface_has_buffer(surface)) {
+		return;
+	}
 
-		struct wlr_subsurface *subsurface;
-		wl_list_for_each(subsurface, &surface->subsurface_list, parent_link) {
-			struct wlr_surface_state *state = subsurface->surface->current;
-			double sx = state->subsurface_position.x;
-			double sy = state->subsurface_position.y;
-			double sw = state->buffer_width / state->scale;
-			double sh = state->buffer_height / state->scale;
-			rotate_child_position(&sx, &sy, sw, sh, width, height, rotation);
+	double sw = surface->current->width * wlr_output->scale;
+	double sh = surface->current->height * wlr_output->scale;
+	double ox = lx, oy = ly;
+	wlr_output_layout_output_coords(desktop->layout, wlr_output, &ox, &oy);
+	ox *= wlr_output->scale;
+	oy *= wlr_output->scale;
+
+	if (wlr_output_layout_intersects(desktop->layout, wlr_output,
+			lx, ly, lx + sw, ly + sh)) {
+		int ow, oh;
+		wlr_output_effective_resolution(wlr_output, &ow, &oh);
+
+		double c = cos(rotation);
+		double s = sin(rotation);
+
+		// Odd transforms are rotations
+		double r = wlr_output->transform % 2 == 0 ? 1 : -1;
+
+		float mat[9] = {
+			+r * (sw * c)/ow, +r * (sh * -s)/ow, r * ((+2.0f * ox + sw) / ow - 1.0f),
+			-r * (sw * s)/oh, -r * (sh * +c)/oh, r * ((-2.0f * oy - sh) / oh + 1.0f),
+			0.0f, 0.0f, 1.0f,
```

Comment:
I preferred this done more explicitly. This isn't really maintainable if we want to change how these matricies are calculated in the future. If it's a performance concern, we can stow the matricies in the view's structure and only recalculate them when the inputs are invalidated.